### ISSUE TITLE
GSDX: Motocross Mania - Bad/Missing Textures + Text

### ISSUE THREAD

**Issue by refractionpcsx2** [PCSX2/pcsx2#1319]
Found this one going through my games checking one of the other issues.  Looks like this has been outstanding for a long time.

Motocross Mania in GSDX, any mode Software or Hardware, OGL/DX.  Fine in ZeroGS.

Here is how it looks in GSDX
![motocrossmania-textures-text](https://cloud.githubusercontent.com/assets/6278726/14898331/1b49b036-0d7e-11e6-8bf2-655b2cae3f91.png)

and here is how it looks in ZeroGS
![mcm-zerogs](https://cloud.githubusercontent.com/assets/6278726/14898350/267944a8-0d7e-11e6-885e-1d48aca0d58b.jpg)

**GS Dump**: [**Here**](https://www.dropbox.com/s/mvlwkygwhld6g0p/MotoCrossMania.rar?dl=0)

**Comment by gregory38** [comment]
Hum, if I enable the dump of input texture in the GSdx SW renderer, I only get strange texture.
Most of them are empty. Some have random value. Some ony contains a single color. I really don't know what going on. Maybe there is a bug in the gif handling stuff but I barely know anything on this stuff.

**Comment by refractionpcsx2** [comment]
Maybe, I dunno, last time I checked it, it looked okay?  unless there is an issue with receiving bits of data.

**Comment by gregory38** [comment]
You mean the gif stuff was ok. Hum, do you know if zzogl is rendering the game correctly (or at least better).

**Comment by refractionpcsx2** [comment]
I could never get zzogl to work for me im afraid, sorry :(

**Comment by gregory38** [comment]
What is the exact game name ?

**Comment by refractionpcsx2** [comment]
Motocross Mania 3

**Comment by gregory38** [comment]
Hum zzogl got the texture right. Unfortunately gs dump aren't 100% compatible between the 2 plugins too bad... Hum there is alpha test mess in the rendering. I'm trying to look at the first letter in the game (checking memory ....). HW renderer give me more info, but I hope I won't have extra issues.

I quickly look at the gif transfer handling it seem to be the same code in both side. Or it is kind of a texture cache issue (there is one to avoid unswizzle in SW). Pallette seem to be fine.

Maybe it could be useful to re-inject GSdx dump in zzogl to have the same testcase.

**Comment by refractionpcsx2** [comment]
Maybe, dunno how you'd do that :P

**Comment by gregory38** [comment]
Actually both format are close. I copy/pasted the code from GSdx to zzogl. I think the major different is the initial save state. Maybe it could work without the load state part. Or maybe I could try to generate 2 dump on the same location. I will try to generate a dump in the first letter screen. Rendering is easier, your scene is more complex. ZZogl shader is more complex, I think the GS memory is uploaded directly to the GPU. Unswizzle is done on the GPU, quite awful to debug.

**Comment by gregory38** [comment]
So I tried to compare 2 dump of the first "paused" screen. But the issue is the initial write of the memory which is wrong. So gsdx dump is already bad. I need to trace all write to the memory which will be harder....

**Comment by gregory38** [comment]
@refractionpcsx2 
I put a log on the gif transfer stuff. First texture write on zzogl. I don't know the unit, likely Bytes.
```
TransferHostLocal start 0x370000 => 0x374000
TransferHostLocal start 0x3f0000 => 0x3f0400
TransferHostLocal start 0x374000 => 0x37c000
TransferHostLocal start 0x3f0800 => 0x3f0c00
TransferHostLocal start 0x37c000 => 0x37d800
TransferHostLocal start 0x3f1800 => 0x3f1c00
```

Then the first write on GSdx (unit is block 256 Bytes)
```
T:Oth	ID:2	S:Info	=> Write! ...  => 0x3f00 W:1 F:C_32 (DIR 00), dPos(0 0) size(64 64)
T:Oth	ID:2	S:Info	=> Write! ...  => 0x3f00 W:1 F:C_32 (DIR 00), dPos(0 0) size(16 16)
T:Oth	ID:2	S:Info	=> Write! ...  => 0x3f08 W:1 F:C_32 (DIR 00), dPos(0 0) size(128 64)
T:Oth	ID:2	S:Info	=> Write! ...  => 0x3f08 W:1 F:C_32 (DIR 00), dPos(0 0) size(16 16)
T:Oth	ID:2	S:Info	=> Write! ...  => 0x3f18 W:1 F:C_32 (DIR 00), dPos(0 0) size(64 16)
T:Oth	ID:2	S:Info	=> Write! ...  => 0x3f18 W:1 F:C_32 (DIR 00), dPos(0 0) size(16 16)
```
I have the feeling that GSdx addresses are potentially wrong.

**Comment by gregory38** [comment]
Hum,  I guess next step is to look at the register content.

**Comment by gregory38** [comment]
So zzogl has a memory effect on the blit buffer. Value is only taken into account after a write to TRX DIR.

ZZogl
```
BLIT 0 13700
TREG 40 40
DIR 0 0
BLIT 0 13f00 // <= here the internal blit buffer is still 0x13700
TransferHostLocal start 0x370000 => 0x374000
TREG 10 10
DIR 0 0
TransferHostLocal start 0x3f0000 => 0x3f0400
```

GSdx
```
BLIT 0 13700
TREG 40 40
DIR 0 0
BLIT 0 13f00
// Value of the transfer is wrong because blit register was updated.
TransferHostLocal start 0x370000 => 0x374000
TREG 10 10
DIR 0 0
TransferHostLocal start 0x3f0000 => 0x3f0400
```

@gabest11 any opinion ? Could it be the core that generate the register write in wrong oder. Or value of blit buffer must be cached.

A fix (i.e. add a buffer to containt transmission parameter) will potentially break all savestate/gsdump :(

**Comment by gregory38** [comment]
I implemented a fix that save a copy of the blit buffer when trxdir is written. Hopefully it doesn't impact save state.

I'm curious if we have a similar issue for others games.

**Comment by refractionpcsx2** [comment]
hmm, so  is the problkem the BLIT write is overwriting the address we do want with a new one? or were we taking the old address with the transfer?

**Comment by gregory38** [comment]
Trxdir ought to start the transfer. But gsdx will do the real memory write later (based on blit buf content).
If you write a new value of blit bif between trxdir and memory write. The latter will use the new value of blit reg (therefore wrong address, wrong format).

In my pr, I just did a copy of the blit reg when trxdir is written. Then in the write I used the "old" copied blit buffer.

Potentially there is a similar issue for the read. And for others transfer registers (like position).
